plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+bindDat :: Datatype TyName Name DefaultUni DefaultFun ()
+        -> GenTm a
+        -> GenTm a
+bindDat dat@(Datatype _ (TyVarDecl _ a k) _ _ _) cont =
+  bindTyName a k $
+  local (\ e -> e { geDatas = Map.insert a dat (geDatas e) }) $
+  foldr (uncurry bindTmName) cont (matchType dat : constrTypes dat)
+
+-- | Bind a binding.
+bindBind :: Binding TyName Name DefaultUni DefaultFun ()
+         -> GenTm a
+         -> GenTm a
+bindBind (DatatypeBind _ dat)              = bindDat dat
+bindBind (TermBind _ _ (VarDecl _ x ty) _) = bindTmName x ty
+-- CODE REVIEW: Should we try to generate type bindings and all the recursive types without datatypes stuff?
+-- I don't think both datatypes and this stuff should actually show up in the same code, no?
+bindBind _                                 = error "unreachable"
+
+-- | Bind multiple bindings
+bindBinds :: Foldable f => f (Binding TyName Name DefaultUni DefaultFun ()) -> GenTm a -> GenTm a
+bindBinds = flip (foldr bindBind)
+
+-- * Generators for well-kinded types
+
+-- | Give a unique "least" (intentionally vaguely specified by "shrinking order")
+-- type of that kind. Note: this function requires care and attention to not get
+-- a shrinking loop. If you think you need to mess with this function:
+-- 1. You're probably wrong, think again and
+-- 2. If you're sure you're not wrong you need to be very careful and
+--    test the shrinking to make sure you don't get in a loop.
+minimalType :: Kind () -> Type TyName DefaultUni ()
+minimalType ty =
+  case ty of
+    Type{} -> unit
+    KindArrow _ k1 k2 ->
+      case k1 : view k2 of
+        [Type{}]         -> list
+        [Type{}, Type{}] -> pair
+        _                -> TyLam () (TyName $ Name "_" (toEnum 0)) k1 $ minimalType k2
+  where
+    view (KindArrow _ k1 k2) = k1 : view k2
+    view _                   = []
+
+    unit = TyBuiltin () (SomeTypeIn DefaultUniUnit)
+    list = TyBuiltin () (SomeTypeIn DefaultUniProtoList)
+    pair = TyBuiltin () (SomeTypeIn DefaultUniProtoPair)
+
+-- | Get the types of builtins at a given kind
+builtinTys :: Kind () -> [SomeTypeIn DefaultUni]
+builtinTys Star =
+  [ SomeTypeIn DefaultUniInteger
+  , SomeTypeIn DefaultUniUnit
+  , SomeTypeIn DefaultUniBool ]
+builtinTys _ = []
+
+-- | Generate "small" types at a given kind such as builtins, bound variables, bound datatypes,
+-- and abstractions /\ t0 ... tn. T
+genAtomicType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genAtomicType k = do
+  tys <- asks geTypes
+  dts <- asks geDatas
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
+      builtins = map (TyBuiltin ()) $ builtinTys k
+      lam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        TyLam () x k1 <$> bindTyName x k1 (genAtomicType k2)
+  oneofTm $ map pure (atoms ++ builtins) ++ [lam k1 k2 | KindArrow _ k1 k2 <- [k]]
```

Comment:
```suggestion
  oneofTm $ map pure (atoms ++ builtins) ++ [lam k1 k2 | let KindArrow _ k1 k2 <- k]
```

Reply:
What syntax extension do I need to turn on to do this?

Reply:
Nothing? I think it's base syntax. or it's in `Haskell2010` which we use everywhere

Reply:
```
λ> :set -XHaskell2010 
λ> [ x | let Just x <- Nothing ]

<interactive>:2:18: error:
    parse error on input ‘<-’
    Perhaps this statement should be within a 'do' block?
```
I've never seen this syntax before in my life...

Reply:
@michaelpj perhaps you mean `let match = k`? But that has a different semantics in a `do`-block. Not sure about comprehensions though, but I'd ban it regardless.

Reply:
Okay, I'm just wrong, it fails with an exception instead of filtering out the case :facepalm: for some reason I thought this worked as a filter.